porch/pkg/apiserver/resolvetoimage.go

```diff
@@ -20,25 +20,53 @@ import (
 	"strings"
 
 	"github.com/GoogleContainerTools/kpt/internal/util/porch"
+	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// KubeFunctionResolver resolves function names to full image paths
+type KubeFunctionResolver struct {
+	client             client.WithWatch
+	defaultImagePrefix string
+	// resolver  *FunctionResolver
+	namespace string
+}
+
 // resolveToImagePorch converts the function short path to the full image url.
 // If the function is Catalog function, it adds "gcr.io/kpt-fn/".e.g. set-namespace:v0.1 --> gcr.io/kpt-fn/set-namespace:v0.1
 // If the function is porch function, it queries porch to get the function image by name and namespace.
 // e.g. default:set-namespace:v0.1 --> us-west1-docker.pkg.dev/cpa-kit-dev/packages/set-namespace:v0.1
-func resolveToImagePorch(ctx context.Context, image string) (string, error) {
+func (r *KubeFunctionResolver) resolveToImagePorch(ctx context.Context, image string) (string, error) {
 	segments := strings.Split(image, ":")
 	if len(segments) == 4 {
 		// Porch function
+		// TODO: Remove this legacy configuration
 		functionName := strings.Join(segments[1:], ":")
 		function, err := porch.FunctionGetter{}.Get(ctx, functionName, segments[0])
 		if err != nil {
-			return "", fmt.Errorf("failed to resolve image: %w", err)
+			return "", fmt.Errorf("failed to get image for function %q: %w", image, err)
 		}
 		return function.Spec.Image, nil
 	}
 	if !strings.Contains(image, "/") {
-		return fmt.Sprintf("gcr.io/kpt-fn/%s", image), nil
+		var function v1alpha1.Function
+		// TODO: Use fieldSelectors and better lookup
+		name := "functions:" + image + ":latest"
+		key := types.NamespacedName{
+			Namespace: r.namespace,
+			Name:      name,
+		}
+		if err := r.client.Get(ctx, key, &function); err != nil {
```

Comment:
Some clarifications:
How's the functions registry API work today ?
I am assuming by default, functions won't be there in any porch installation, so this resolving should be opt-in. Simplify because it is now calling network call for every function with out `/`. WDYT ?

Reply:
How it works today: You can register an OCI Repository with `spec.content: Function`, and then any discovered images are registered as Functions.

Performance: Currently I don't _think_ the client.Client is caching (but it's always hard to know with controller-runtime!), but we could make it caching if we're concerned about the performance.

Opt-in: I think this should be the default behaviour and the default way we use functions.  It's somewhat akin to the Function Catalog idea from kustomize.  That said, I expect it to evolve over time (not least because we want to do something on the client side as well).  If we have a flag and we set it for users then everyone will be running with anyway :-)  I think we have a degree of opt-in because the user has to register a Repository of spec.content=Function to activate the behaviour (otherwise we fall through).

I do think I should write a doc about how this works today, we can keep it updated even as it changes...

Reply:
I think `function registry` is not widely discussed or shown in any of the demos, so I always considered as `aspirational` :) but this discussion suggests it is not the case.

Re: `performance`, I think `functions` are always beaten for slow performance, so just being careful with adding new latencies :)

Yes, I think documentation will certainly help.

For now, let's go with this and see if it causes any trouble in the field.